File: src/calcs/empirical.js
import { prepInputVectors, roundValueFixed } from "./utils"

var nj = require("jsnumpy")
var linterp = require("everpolate").linear

// https://gist.github.com/janosh/099bd8061f15e3fbfcc19be0e6b670b9
const argFact = (compareFn) => (array) => array.map((el, idx) => [el, idx]).reduce(compareFn)[1];
const argMin = argFact((max, el) => (el[0] < max[0] ? el : max));

function getCDF(yVector) {
    let countY = yVector.length;
    let listArea = nj.fillWithNumber([countY]);
    listArea[0] = 0;
    let cumArea = 0;
    for (let i = 0; i < countY-1; i++) {
        cumArea = cumArea + (yVector[i+1] + yVector[i]) / 2 / (countY - 1);
        listArea[i+1] = cumArea;
    }
    return nj.divide([listArea], cumArea)[0];
}

function getMean(xVector, yVector) {
    let mean = 0;
    let countY = yVector.length;
    for (let i = 0; i < countY-1; i++) { 
        mean = mean + (xVector[i+1] + xVector[i]) / 2 * (yVector[i+1] + yVector[i]) / 2 / (countY - 1);
    }
    return mean; 
}

function getMedian(xVector, cdf) {
    return xVector[argMin(nj.abs(nj.subtract([cdf], 0.5))[0])];
}

function getStdDev(xVector, yVector) {
    return (getMean(nj.power(xVector, 2), yVector) - getMean(xVector, yVector) ** 2) ** 0.5;
}

function getStats(xCoords, yCoords, xMin, xMax) {
    let xVector = prepInputVectors(xCoords, yCoords, xMin, xMax).x;
    let yVector = prepInputVectors(xCoords, yCoords, xMin, xMax).y;

    let cdf    = getCDF(yVector);
    
    let mean   = getMean(xVector, yVector);
    let median = getMedian(xVector, cdf);
    let std    = getStdDev(xVector, yVector);

    return {
            mean   : roundValueFixed(mean, xMin, xMax, 5),
            median : roundValueFixed(median, xMin, xMax, 5),
            std    : roundValueFixed(std, xMin, xMax, 5)
           };
}

function getSamples(xCoords, yCoords, xMin, xMax, nSamples) {
    if (xCoords.length <=1 ) {
        return [];
    }
    let xVector = prepInputVectors(xCoords, yCoords, xMin, xMax).x;
    let yVector = prepInputVectors(xCoords, yCoords, xMin, xMax).y;
    let cdf = getCDF(yVector);
    let unifSamples = Array.from({length: nSamples}, () => Math.random());

    let samples = linterp(unifSamples, cdf, xVector);
    for (let i = 0; i < samples.length; i++) {
        samples[i] = roundValueFixed(samples[i], xMin, xMax, 5);
    }
    return samples;
}

export { getStats, getSamples }

File: api/empirical.py
import numpy as np
from utils import prep_input_vectors


def get_cdf(y_vector: np.ndarray) -> np.ndarray:
    '''
    Get the CDF from a y vector (assumes interpolated x)

    Args:
        y_vector: The y vector to get the CDF of.
    Returns:
        np.ndarray: The resulting CDF.

    Examples:
        >>> y_vec = np.array([1, 2, 3, 4, 5])
        >>> cdf = get_cdf(y_vector)
    '''
    count_y = len(y_vector)
    list_area = np.zeros(count_y)
    cum_area = 0
    for i in range(count_y - 1):
        cum_area += (y_vector[i+1] + y_vector[i]) / 2 / (count_y - 1)
        list_area[i+1] = cum_area
    return (list_area) / np.max(list_area)

def get_mean(x_vector, y_vector):
    '''
    Calculates the mean of the distribution that the x and y vectors represent. Assumes that the x vector has been interpolated.
    '''
    mean = 0
    count_y = len(y_vector)
    for i in range(count_y - 1):
        mean += (x_vector[i+1] + x_vector[i]) / 2 * (y_vector[i+1] + y_vector[i]) / 2 / (count_y - 1)
    return mean

def get_median(x_vector, cdf):
    return x_vector[np.argmin(np.abs(cdf - 0.5))]

def get_std(x_vector, y_vector):
    return (get_mean(x_vector ** 2, y_vector) - get_mean(x_vector, y_vector) ** 2) ** 0.5

def get_stats(
              x_vector: np.ndarray,
              y_vector: np.ndarray,
              ) -> dict:
    '''
    Given a y vector, an x vector and a CDF use a provided x vector range to 
    extract the mean, median and standard deviation of the distribution that they represent.

    Args:
        x_vector: The vector of x points.
        y_vector: The vector of y points.
    Returns:
        tuple: The mean, median and standard deviation of the distribution that the x and y vectors represent.
    '''
    cdf = get_cdf(y_vector)
    mean = get_mean(x_vector, y_vector)
    median = get_median(x_vector, cdf)
    std = get_std(x_vector, y_vector)
    return {
            'mean': mean,
            'median': median,
            'std': std
           }

def get_samples(x_coords, y_coords, x_min, x_max, n_samples):
    '''
    Given a set of x and y coordinates, a minimum and maximum x value and a number of samples,
    this function returns a set of samples from the distribution represented by the x and y coordinates.

    Args:
        x_coords: The x coordinates of the distribution.
        y_coords: The y coordinates of the distribution.
        x_min: The minimum x value of the distribution.
        x_max: The maximum x value of the distribution.
        n_samples: The number of samples to generate.
    Returns:
        np.ndarray: The x and y coordinates of the samples.
    '''
    x_vector, y_vector = prep_input_vectors(x_coords, y_coords, x_min, x_max)
    cdf = get_cdf(y_vector)
    unif_samples = np.random.uniform(0, 1, n_samples)
    samples = np.interp(unif_samples, cdf, x_vector)
    return samples